### ISSUE TITLE
Slow binding of PV => PVC

### ISSUE THREAD

**Issue by qix** [kubernetes/kubernetes#21498]
We're manually creating drives to contain each of our databases, as such it makes sense for us to have known bindings between each PVC and PV. Since `volumeSelector` is not yet available creating them one at a time has been working for us.

Recently for unknown reasons it has gotten **very** slow. What used to take seconds to bind the PVC=>PV now takes to 20+ minutes.

Initially it shows up as `Pending` with errors such as the following in the logs:
`PVClaimBinder 
could not update claim sliding-data-3c-pvc: Error updating volume: persistentvolumeclaims "sl
iding-data-3c-pvc" cannot be updated: the object has been modified; please apply your changes
 to the latest version and try again`

After a while it finally binds, but with `Capacity: 0`:
```
# kubectl describe pv sliding-data-1c-pv
Name:        sliding-data-1c-pv
Labels:        <none>
Status:        Available
Claim:        
Reclaim Policy:    Retain
Access Modes:    RWO
Capacity:    16Gi
Message:    
Source:
    Type:    GCEPersistentDisk (a Persistent Disk resource in Google Compute Engine)
    PDName:    sliding-data-1c
    FSType:    ext4
    Partition:    0
    ReadOnly:    false

# kubectl describe pvc sliding-data-1c-pvc
Name:        sliding-data-1c-pvc
Namespace:    default
Status:        Bound
Volume:        sliding-data-1c-pv
Labels:        <none>
Capacity:    0
Access Modes:    
```

Finally after around half an hour everything settles down as expected. Since we can only create one at a time, this means many hours while trying to create a group of PV/PVC pairs.

We're running our own Kubernetes cluster on Google Compute Engine, here are some other details:

```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"1", GitVersion:"v1.1.3", GitCommit:"6a81b50c7e97bbe0ade075de55ab4fa34f049dc2", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"1", GitVersion:"v1.1.3", GitCommit:"6a81b50c7e97bbe0ade075de55ab4fa34f049dc2", GitTreeState:"clean"}

# cat sliding-data-3c-pv.json
{
  "apiVersion": "v1",
  "kind": "PersistentVolume",
  "metadata": {
    "name": "sliding-data-3c-pv"
  },
  "spec": {
    "accessModes": [
      "ReadWriteOnce"
    ],
    "capacity": {
      "storage": "16Gi"
    },
    "gcePersistentDisk": {
      "fsType": "ext4",
      "pdName": "sliding-data-3c"
    },
    "persistentVolumeReclaimPolicy": "Retain"
  }
}

# cat sliding-data-3c-pvc.json 
{
  "apiVersion": "v1",
  "kind": "PersistentVolumeClaim",
  "metadata": {
    "name": "sliding-data-3c-pvc"
  },
  "spec": {
    "accessModes": [
      "ReadWriteOnce"
    ],
    "resources": {
      "requests": {
        "storage": "16Gi"
      }
    }
  }
}

# docker --version
Docker version 1.6.2, build 7c8fca2
```

**Comment by saad-ali** [comment]
CC @kubernetes/sig-storage @kubernetes/rh-storage

**Comment by jsafrane** [comment]
The main reason is long periodical update of clam/PV sync loop (10 minutes by default). Try `kube-controller-manager --pvclaimbinder-sync-period=15s` or even less. Everything should settle after 30 seconds in the worst case.

Also, PVCs should be bound immediately if you create PV first and PVC second.

We're working on improving the binder to remove any unnecessary delays, but it will take time.

**Comment by saad-ali** [comment]
Expanding on @jsafrane's suggestion. Here is an example showing how to pre-bind volumes:

Create a PV object with a ClaimRef field referencing a PVC that you will subsequently create:
```BASH
$ kubectl create -f pv.yaml 
persistentvolume "pv0003" created
```
where `pv.yaml` contains:
```YAML
  apiVersion: v1
  kind: PersistentVolume
  metadata:
    name: pv0003
  spec:
    capacity:
      storage: 5Gi
    accessModes:
      - ReadWriteOnce
    persistentVolumeReclaimPolicy: Recycle
    claimRef:
      namespace: default
      name: myclaim
    nfs:
      path: /tmp
      server: 172.17.0.2
```
Then create the PVC:
```YAML
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: myclaim
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
```
The PV and PVC should be bound immediately:
```BASH
$ kubectl get pvc
NAME      STATUS    VOLUME    CAPACITY   ACCESSMODES   AGE
myclaim   Bound     pv0003    5Gi        RWO           4s
$ ./cluster/kubectl.sh get pv
NAME      CAPACITY   ACCESSMODES   STATUS    CLAIM             REASON    AGE
pv0003    5Gi        RWO           Bound     default/myclaim             57s
```

**Comment by markturansky** [comment]
@saad-ali the UID must also match, else it might not be the same claim:  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/persistentvolume/types.go#L120

**Comment by saad-ali** [comment]
> @saad-ali the UID must also match, else it might not be the same claim: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/persistentvolume/types.go#L120

Great point Mark. Updated the example to specify UID.

**Comment by erinboyd** [comment]
Interesting to manually specify the UID in the PV...Curious why this is necessary though. If you have the claimRef and name and you are respecting those, why wouldn't you just reduce the latency and still use the system generated UID from the PVC to populate the PV. What is the need for the 2 step verification here? Is it the generating of the UID that is causing the latency? 

As far as the latency goes, this also touches on the bootstrapping case where we pre-bind and would possibly time out with ansible waiting on the bind to happen so I think this is a great idea.

Note the example won't work with recycling though, as once it's created it deletes the claimRef.

**Comment by erinboyd** [comment]
adding to types.go
   // do one more search just for name
            if claim.Name == volume.Spec.ClaimRef.Name && claim.Namespace == volume.Spec.ClaimRef.Namespace {

Will make sure the bind is respected by the name then once the pvc is created and the UID generated it can be populated back into the PV. Reducing the need for the user to supply both.

**Comment by erinboyd** [comment]
Some of this is repeat of the discussion in https://github.com/kubernetes/kubernetes/issues/23615

**Comment by saad-ali** [comment]
>> @saad-ali the UID must also match, else it might not be the same claim: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/persistentvolume/types.go#L120
>
> Great point Mark. Updated the example to specify UID.

.
> Interesting to manually specify the UID in the PV.

Thinking about this some more, the user **should not** specify the UID, and, in fact, if one is specified, it is ignored (thanks to @thockin for pointing that out). The existing code allows you to prebind a PV to PVC that doesn't exist yet by creating the PV with a claimRef including just the unique namespace/name. Then as long as a PVC with the same namespace/name is created pre-binding should just work. The existing design may have some edge cases that don't work quite right. The redesign of this logic aims to continue to support this behavior so it should continue to work and become more robust in the future. I'll modify the example above to remove references to UID.

**Comment by markturansky** [comment]
@saad-ali I agree name/namespace is likely sufficient to pre-bind to a PVC.  The UID was added to solve a corner case, though.

Consider:

1. Create PV named "foo"
2. Create PVC named "bar" that matches "foo". Let it bind.
3. Delete the PVC.  Now we race.
3. Create a new PVC named "bar".

The volume should be released of its first claim.  Without the UID check, the PV can incorrectly bind to the 2nd claim, which is a different object.

**Comment by thockin** [comment]
The UID is (will be) applied when the binding is actually made by the
controller.  The user is requesting a binding, but the controller needs to
ACK it.

**Comment by erinboyd** [comment]
but I think the argument exists if the user can create a UID of their own...Seems like something you would then have to prevalidate that it doesn't already exist. So the expense of this operation alone, I think outweighs the use of doing this. Especially if the UID isn't populated until the bind is successful.

**Comment by thockin** [comment]
Erin,

I don't follow.  Users can not create a UID of their own.  Never could,
never will.

**Comment by erinboyd** [comment]
@thockin Yes, I am aware. Mark was suggesting it above and I was saying I didn't think we should consider it

**Comment by markturansky** [comment]
@erinboyd I said that the UID is required on the ClaimRef to be a valid bind, not that the user will supply a UID.

**Comment by erinboyd** [comment]
@markturansky my bad. ;)

**Comment by jeffvance** [comment]
I have a spreadsheet that summarizes some of the slowness: https://docs.google.com/spreadsheets/d/12XPwt53ueqpP2z1HyAuLdKjLmCOsGtqn59OSBJPY0Zk/edit#gid=484771528
Summary is that when a claim is created before its matching pv, the binding can take over 500x longer than if the pv is created first. 
I experimented with various pvclaimbinder-sync-period values which are shown in the spreadsheet. Using 15s vs the default 10m reduces binding time (for the pvc created first) up to 45x!
See also PR #24682.

**Comment by erinboyd** [comment]
@jsafrane please see findings in spreadsheet.

**Comment by yunjing** [comment]
@saad-ali Thanks for speaking with @qix and I (from smyte) the other day and suggesting the pre-binding option. Here are a few things we found:
1. Pre-binding only works after we upgraded to v1.2, which is also suggested in this commit https://github.com/kubernetes/kubernetes/commit/b9b8cf7fff297df724d9236240e84ba1fd97537e.
2. At v1.2.2, PV binding is still taking a few minutes even though we set `pvclaimbinder-sync-period=15s`.
3. As before, restarting k8s-controller helped accelerate the binding process.

Practically speaking, since we can now bind many PVs at the same time, taking a few minutes to bind is not a primary concern anymore.

**Comment by saad-ali** [comment]
@yunjing Glad to hear you guys are unblocked. There is a big redesign underway (see https://github.com/kubernetes/kubernetes/pull/24331) for PV/PVC binding code that should fix these latency in binding issues. You can expect that in v1.3.

**Comment by childsb** [comment]
Has anyone tested binding latency on kube 1.3?  I'd like to close this!

**Comment by thockin** [comment]
Not joking, we should bulk close all volume-controller, binder, and
attacher bugs that predate 1.3 - all of that has been rewritten.

Volume driver bugs can stay, of course ;)

**Comment by matchstick** [comment]
We believe this is addressed in 1.3.4 Please reopen if it reoccured.
